diff --git a/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java b/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
index dae940982a..e06ab44593 100644
--- a/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
+++ b/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
@@ -527,19 +527,29 @@ private boolean expandSelectItem(final SqlNode selectItem, SqlSelect select,
   }
 
   private static SqlNode expandExprFromJoin(SqlJoin join,
-      SqlIdentifier identifier, SelectScope scope) {
-    if (join.getConditionType() != JoinConditionType.USING) {
+      SqlIdentifier identifier, SelectScope scope, SqlValidatorImpl validator) {
+    if (join.getConditionType() != JoinConditionType.USING && !join.isNatural()) {
       return identifier;
     }
+    List<String> commonColumnNames;
+    // must be natural or using here, and cannot specify NATURAL keyword with USING clause
+    if (join.isNatural()) {
+      commonColumnNames = validator.deriveNaturalJoinColumnList(join);
+    } else {
+      commonColumnNames = SqlIdentifier.simpleNames((SqlNodeList) getCondition(join));
+    }
 
+    final boolean caseSensitive = validator.catalogReader.nameMatcher().isCaseSensitive();
     final Map<String, String> fieldAliases = getFieldAliases(scope);
 
-    for (String name
-        : SqlIdentifier.simpleNames((SqlNodeList) getCondition(join))) {
-      if (identifier.getSimple().equals(name)) {
+    for (String name : commonColumnNames) {
+      if (Util.matches(caseSensitive, identifier.getSimple(), name)) {
         final List<SqlNode> qualifiedNode = new ArrayList<>();
         for (ScopeChild child : requireNonNull(scope, "scope").children) {
-          if (child.namespace.getRowType().getFieldNames().contains(name)) {
+          if (Util.findMatch(
+               child.namespace.getRowType().getFieldNames(),
+               name,
+               caseSensitive) > -1) {
             final SqlIdentifier exp =
                 new SqlIdentifier(
                     ImmutableList.of(child.name, name),
@@ -561,7 +571,7 @@ private static SqlNode expandExprFromJoin(SqlJoin join,
           return coalesceCall;
         } else {
           return SqlStdOperatorTable.AS.createCall(SqlParserPos.ZERO, coalesceCall,
-              new SqlIdentifier(name, SqlParserPos.ZERO));
+              new SqlIdentifier(identifier.getSimple(), SqlParserPos.ZERO));
         }
       }
     }
@@ -570,7 +580,7 @@ private static SqlNode expandExprFromJoin(SqlJoin join,
     // since it is always left-deep join.
     final SqlNode node = join.getLeft();
     if (node instanceof SqlJoin) {
-      return expandExprFromJoin((SqlJoin) node, identifier, scope);
+      return expandExprFromJoin((SqlJoin) node, identifier, scope, validator);
     } else {
       return identifier;
     }
@@ -642,7 +652,10 @@ private static SqlNode expandCommonColumn(SqlSelect sqlSelect,
       return selectItem;
     }
 
-    return expandExprFromJoin((SqlJoin) from, identifier, scope);
+    return expandExprFromJoin((SqlJoin) from,
+         identifier,
+         scope,
+         validator);
   }
 
   private static void validateQualifiedCommonColumn(SqlJoin join,
diff --git a/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java b/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
index ccbd447ac7..2c3ff38ba5 100644
--- a/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
+++ b/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
@@ -5466,6 +5466,70 @@ private ImmutableList<ImmutableBitSet> cube(ImmutableBitSet... sets) {
         .ok();
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-7051">[CALCITE-7051]
+   * JOIN with USING does not match the appropriate columns when caseSensitive is false</a>. */
+  @Test void testSelectJoinUsingCommonColumnCaseSensitive() {
+
+    sql("select DEPTNO from emp join dept using (DEPTNO)")
+        .withCaseSensitive(false)
+        .withValidatorIdentifierExpansion(true)
+        .withQuotedCasing(Casing.UNCHANGED)
+        .withUnquotedCasing(Casing.UNCHANGED)
+        .ok();
+
+    sql("select deptno from emp join dept using (deptno)")
+        .withCaseSensitive(false)
+        .withValidatorIdentifierExpansion(true)
+        .withQuotedCasing(Casing.UNCHANGED)
+        .withUnquotedCasing(Casing.UNCHANGED)
+        .ok();
+
+    sql("select DEPTNO from emp join dept using (deptno)")
+        .withCaseSensitive(false)
+        .withValidatorIdentifierExpansion(true)
+        .withQuotedCasing(Casing.UNCHANGED)
+        .withUnquotedCasing(Casing.UNCHANGED)
+        .ok();
+
+    sql("select deptno from emp join dept using (DEPTNO)")
+        .withCaseSensitive(false)
+        .withValidatorIdentifierExpansion(true)
+        .withQuotedCasing(Casing.UNCHANGED)
+        .withUnquotedCasing(Casing.UNCHANGED)
+        .ok();
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-7051">[CALCITE-7051]
+   * JOIN with USING does not match the appropriate columns when caseSensitive is false</a>.
+   * select common columns in NATURAL JOIN*/
+  @Test void testNaturalJoinCommonColumn() {
+    sql("select DEPTNO from emp natural join dept")
+        .withCaseSensitive(true)
+        .withValidatorIdentifierExpansion(true)
+        .ok();
+
+    sql("select deptno from emp natural join dept")
+        .withCaseSensitive(true)
+        .withValidatorIdentifierExpansion(true)
+        .ok();
+
+    sql("select deptno from emp natural join dept")
+        .withCaseSensitive(false)
+        .withValidatorIdentifierExpansion(true)
+        .withQuotedCasing(Casing.UNCHANGED)
+        .withUnquotedCasing(Casing.UNCHANGED)
+        .ok();
+
+    sql("select DEPTNO from emp natural join dept")
+        .withCaseSensitive(false)
+        .withValidatorIdentifierExpansion(true)
+        .withQuotedCasing(Casing.UNCHANGED)
+        .withUnquotedCasing(Casing.UNCHANGED)
+        .ok();
+  }
+
   @Test void testCrossJoinOnFails() {
     sql("select * from emp cross join dept\n"
         + " ^on^ emp.deptno = dept.deptno")
